modules/basegl/src/processors/layerprocessing/layerglprocessor.cpp

```diff
@@ -0,0 +1,115 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/basegl/processors/layerprocessing/layerglprocessor.h>
+#include <inviwo/core/datastructures/image/layer.h>
+#include <inviwo/core/datastructures/image/imagetypes.h>
+#include <modules/opengl/image/layergl.h>
+#include <modules/opengl/shader/shader.h>
+#include <modules/opengl/shader/shaderutils.h>
+#include <modules/opengl/shader/standardshaders.h>
+#include <modules/opengl/texture/textureunit.h>
+#include <modules/opengl/texture/textureutils.h>
+
+namespace inviwo {
+
+LayerGLProcessor::LayerGLProcessor(std::shared_ptr<const ShaderResource> fragmentShader,
+                                   Shader::Build build)
+    : Processor()
+    , inport_("inport", "The input layer"_help)
+    , outport_("output", "Resulting output layer"_help)
+    , dataFormat_(nullptr)
+    , swizzleMask_(swizzlemasks::rgba)
+    , internalInvalid_(false)
+    , shader_({utilgl::imgQuadVert(), {ShaderType::Fragment, fragmentShader}}, build) {
+
+    addPorts(inport_, outport_);
+
+    inport_.onChange([this]() {
+        markInvalid();
+        afterInportChanged();
+    });
+
+    shader_.onReload([this]() { invalidate(InvalidationLevel::InvalidResources); });
+}
+
+void LayerGLProcessor::initializeResources() {
+    shader_.build();
+    internalInvalid_ = true;
+}
+
+void LayerGLProcessor::process() {
+    bool reattach = false;
+
+    if (internalInvalid_) {
+        reattach = true;
+        internalInvalid_ = false;
+        layer_ = std::make_shared<Layer>(*inport_.getData(), noData, std::nullopt, dataFormat_);
+        if (dataFormat_) {
+            layer_->setSwizzleMask(swizzleMask_);
+        }
```

Comment:
why if(dataFormat)?

Reply:
So that swizzleing is only applied if there is a custom output format defined.